src/EventStore.Core/TransactionLog/Chunks/TFChunkScavenger.cs

```diff
@@ -23,60 +23,133 @@ public class TFChunkScavenger
         private static readonly ILogger Log = LogManager.GetLoggerFor<TFChunkScavenger>();
 
         private readonly TFChunkDb _db;
-        private readonly IODispatcher _ioDispatcher;
+        private readonly ITFChunkScavengerLog _scavengerLog;
         private readonly ITableIndex _tableIndex;
-        private readonly Guid _scavengeId;
-        private readonly string _nodeEndpoint;
         private readonly IReadIndex _readIndex;
         private readonly long _maxChunkDataSize;
         private readonly bool _unsafeIgnoreHardDeletes;
+        private readonly int _threads;
         private const int MaxRetryCount = 5;
+        internal const int MaxThreadCount = 4;
 
-        public TFChunkScavenger(TFChunkDb db, IODispatcher ioDispatcher, ITableIndex tableIndex, IReadIndex readIndex,
-                                Guid scavengeId, string nodeEndpoint, long? maxChunkDataSize = null, bool unsafeIgnoreHardDeletes=false)
+        public TFChunkScavenger(TFChunkDb db, ITFChunkScavengerLog scavengerLog, ITableIndex tableIndex, IReadIndex readIndex, long? maxChunkDataSize = null,
+            bool unsafeIgnoreHardDeletes = false, int threads = 1)
         {
             Ensure.NotNull(db, "db");
-            Ensure.NotNull(ioDispatcher, "ioDispatcher");
+            Ensure.NotNull(scavengerLog, "scavengerLog");
             Ensure.NotNull(tableIndex, "tableIndex");
-            Ensure.NotNull(nodeEndpoint, "nodeEndpoint");
             Ensure.NotNull(readIndex, "readIndex");
+            Ensure.Positive(threads, "threads");
+
+            if (threads > MaxThreadCount)
+            {
+                Log.Warn("{0} scavenging threads not allowed.  Max threads allowed for scavenging is {1}. Capping.", threads, MaxThreadCount);
+                threads = MaxThreadCount;
+            }
 
             _db = db;
-            _ioDispatcher = ioDispatcher;
+            _scavengerLog = scavengerLog;
             _tableIndex = tableIndex;
-            _scavengeId = scavengeId;
-            _nodeEndpoint = nodeEndpoint;
             _readIndex = readIndex;
             _maxChunkDataSize = maxChunkDataSize ?? db.Config.ChunkSize;
             _unsafeIgnoreHardDeletes = unsafeIgnoreHardDeletes;
+            _threads = threads;
         }
 
-        public long Scavenge(bool alwaysKeepScavenged, bool mergeChunks)
-        {
-            var totalSw = Stopwatch.StartNew();
-            var sw = Stopwatch.StartNew();
-            long spaceSaved = 0;
-
-            Log.Trace("SCAVENGING: started scavenging of DB. Chunks count at start: {0}. Options: alwaysKeepScavenged = {1}, mergeChunks = {2}",
-                      _db.Manager.ChunksCount, alwaysKeepScavenged, mergeChunks);
+        public string ScavengeId => _scavengerLog.ScavengeId;
 
-            for (long scavengePos = 0; scavengePos < _db.Config.ChaserCheckpoint.Read(); )
+        private IEnumerable<TFChunk.TFChunk> GetAllChunks(int startFromChunk)
```

Comment:
private IEnumerable<TFChunk.TFChunk> GetAllChunks(int startFromChunk)
        {
            for (int chunkNum = startFromChunk; chunkNum < _db.Manager.ChunksCount; )
            {
                var chunk = _db.Manager.GetChunk(chunkNum);
                if (!chunk.IsReadOnly)
                {
                    yield break;
                }

                yield return chunk;
                
                chunkNum = chunk.ChunkHeader.ChunkEndNumber + 1;
            }
        }

Reply:
I'm not sure how this doesn't work the way it is. Can you provide a failing test or explain the problem?

Reply:
Your starting `scavengePos`, since it uses a fixed `_db.Config.ChunkSize` assumes chunk sizes are constant in relation to the checkpoint position. They aren’t once scavenged.

Reply:
Well whatever it is, it didn’t work on my test database.